radix-engine/src/transaction/transaction_receipt.rs

```diff
@@ -88,6 +221,10 @@ pub struct TransactionFeeDetails {
     pub execution_cost_breakdown: BTreeMap<String, u32>,
     /// Finalization cost breakdown
     pub finalization_cost_breakdown: BTreeMap<String, u32>,
+    // TODO: This is a breaking change to the receipt and will be fixed in the future when we add
```

Comment:
Bringing this to the attention of the reviewers

Reply:
Here is what I imagined how the DTO thing would work.

``` rust
VersionedTransactionReceipt {
    V1(TransactionReceiptV1) // DTO, identical to the current TransactionReceipt
    V2(TransactionReceiptV2) // another version if any
}
```

The above will be used by Wallet, Node and RET for compatibility. The `TransactionReceipt` in scrypto repo will be used for runtime only, e.g. being used by Node for mapping into receipt DTO or by Scrypto tests.

With this in mind, would it make sense to just mark the field as `skipped` so no compatibility-breaking change is introduced, before DTO is implemented.

Reply:
@iamyulong this can't be used for the wallet, because you might have a wallet which only understands V1 talking to a node outputting V2. 

For now, we need to always return `VersionedTransactionReceipt::V1` from the API.

I suggest a DTO strategy for the preview will be preview-specific and be something with a permissive parser which supports additions (such as JSON or protocol buffers).

Reply:
Using `#[sbor(skip)]` would also work! I forgot that that exists. I think it's possible the first concrete use case for it I've seen aha.

Reply:
Thank you both for your input on this! I think we can just treat the way that the information is exposed in this PR as being temporary since I really just needed a way to get the information from the costing module and into the tests that consume the information. So, we can think of a better way to expose this information.

> We can't do this. For the time being, for wallet compatibility, we need TransactionReceiptV1 to be fixed.

Yup, agreed, this was more of a way to expose this information in some way until we agree on a proper way of exposing it.


> The DTO will need to be additional to the TransactionReceiptV1 until the wallet can migrate to the DTO and enough time has past that we have enough people using wallets with the DTO that we can remove the transaction receipt.

Yup, agreed. I share similar thoughts on this. Will outline some my thoughts on this at the end of this comment.

> With this in mind, would it make sense to just mark the field as skipped so no compatibility-breaking change is introduced, from persistence, encoding and decoding perspective, before DTO is implemented.

I like this! This would give us a way to do this before implementing the DTOs. I guess we would then need to make this field optional (not an issue, just more of a question)? I guess I'm not entirely sure what `#[sbor(skip)]` does to be able to reason about this.

---

I have a small proposal on how we can do the whole receipt versioning primarily with respect to the wallet.

1. We would continue to define `VersionedTransactionReceipt` as an enum of versions V1 all the way up to VN.
2. V(N+1) can always be downgraded to VN and VN can always be upgraded to V(N+1).
3. The core api and gateway api accept an additional argument in transaction preview of the version of receipt desired by the wallet.
4. Each wallet version knows the version of receipt it's compatible with and requests that from the gateway when doing a preview.
5. CoreAPI and Gateway would no longer return `VersionedTransactionReceipt` but would return `TransactionReceiptVX` where X is a version specified by the client in the request.

I see this as being a very simple way to go about this. Thoughts?

Reply:
> @iamyulong this can't be used for the wallet, because you might have a wallet which only understands VersionedTransactionReceipt::V1 talking to a node outputting latest = VersionedTransactionReceipt::V2.

This is what typically wallet deprecation solves. In the future, you may not be able to convert V2 into a V2 for legacy clients.

Reply:
Thank you all for this discussion! I will mark this as resolved since we've already discussed it on slack.